- ### Titanic Dataset - Kaggle

- ### Objetivos
    - #### Explorar os dados
    - #### Pré-processamento dos dados
    - #### Análise exploratória de dados (EDA)
    - #### Treinar um modelo

In [8]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [9]:
dataset = pd.read_csv('train.csv')
dataset.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [10]:
dataset.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [11]:
passageiros_classe_3 = dataset.loc[dataset['Pclass'] == 3]
passageiros_classe_3

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
882  female  22.0      0      0              7552  10.5167   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[491 rows x 12 columns]

In [12]:
classe_cabine_3 = dataset[dataset['Pclass'] == 3]['Cabin']
classe_cabine_3

0      NaN
2      NaN
4      NaN
5      NaN
7      NaN
      ... 
882    NaN
884    NaN
885    NaN
888    NaN
890    NaN
Name: Cabin, Length: 491, dtype: object

In [13]:
# Analise para classe 2

passageiros_classe_2 = dataset.loc[dataset['Pclass'] == 2]
passageiros_classe_2

PassengerId  Survived  Pclass  \
9             10         1       2   
15            16         1       2   
17            18         1       2   
20            21         0       2   
21            22         1       2   
..           ...       ...     ...   
866          867         1       2   
874          875         1       2   
880          881         1       2   
883          884         0       2   
886          887         0       2   

                                             Name     Sex   Age  SibSp  Parch  \
9             Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1      0   
15               Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0      0   
17                   Williams, Mr. Charles Eugene    male   NaN      0      0   
20                           Fynney, Mr. Joseph J    male  35.0      0      0   
21                          Beesley, Mr. Lawrence    male  34.0      0      0   
..                                            ...     ...   ...    ...    ...   
866                  Duran y More, Miss. Asuncion  female  27.0      1      0   
874         Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1      0   
880  Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0      1   
883                 Banfield, Mr. Frederick James    male  28.0      0      0   
886                         Montvila, Rev. Juozas    male  27.0      0      0   

               Ticket     Fare Cabin Embarked  
9              237736  30.0708   NaN        C  
15             248706  16.0000   NaN        S  
17             244373  13.0000   NaN        S  
20             239865  26.0000   NaN        S  
21             248698  13.0000   D56        S  
..                ...      ...   ...      ...  
866     SC/PARIS 2149  13.8583   NaN        C  
874         P/PP 3381  24.0000   NaN        C  
880            230433  26.0000   NaN        S  
883  C.A./SOTON 34068  10.5000   NaN        S  
886            211536  13.0000   NaN        S  

[184 rows x 12 columns]

In [14]:
classe_cabine_2 = dataset[dataset['Pclass'] == 2]['Cabin']
classe_cabine_2

9      NaN
15     NaN
17     NaN
20     NaN
21     D56
      ... 
866    NaN
874    NaN
880    NaN
883    NaN
886    NaN
Name: Cabin, Length: 184, dtype: object

In [15]:
# Analise para classe 1

passageiros_classe_1 = dataset.loc[dataset['Pclass'] == 1]
passageiros_classe_1

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
11            12         1       1   
23            24         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
11       0    113783  26.5500         C103        S  
23       0    113788  35.5000           A6        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[216 rows x 12 columns]

In [16]:
classe_cabine_1 = dataset[dataset['Pclass'] == 1]['Cabin']
classe_cabine_1

1              C85
3             C123
6              E46
11            C103
23              A6
          ...     
871            D35
872    B51 B53 B55
879            C50
887            B42
889           C148
Name: Cabin, Length: 216, dtype: object

In [17]:
print('Valores NaN na classe 1:',classe_cabine_1.isnull().sum())
print('Valores NaN na classe 2:',classe_cabine_2.isnull().sum())
print('Valores NaN na classe 3:',classe_cabine_3.isnull().sum())

Valores NaN na classe 1: 40
Valores NaN na classe 2: 168
Valores NaN na classe 3: 479


- #### Neste caso, vemos como os dados NaN podem ser úteis. Com eles, conseguimos notar que os passageiros da classe 3 tem a maioria dos valores NaN, o que sugere que muitos passageiros dessa classe não tinham uma cabine especificamente atribuída.

In [18]:
agrupado = dataset.groupby(['Embarked']).size()
agrupado

Embarked
C    168
Q     77
S    644
dtype: int64

- #### "C": Cherbourg, França - 168 passageiros.
- #### "Q": Queenstown, Irlanda - 77 passageiros.
- #### "S": Southampton, Inglaterra - 644 passageiros.

- ### Treinar um modelo utilizando arvores de decisão

In [21]:
dataset.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [58]:
# Separar entre sexo, idade e classe X
# Sobreviveu Y

x = dataset[['Sex', 'Age', 'Pclass']].copy()
y = dataset['Survived']

# Na coluna 'Sex', colocaremos 0 para masculino e 1 para feminino

x['Sex'] = x['Sex'].map({"male": 0, "female": 1})

# A arvore vai tratar como um valor diferente de todos os outros
x = x.fillna(-1)

In [70]:
xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size = 0.3, random_state = 42)

modelo = DecisionTreeClassifier(random_state = 0, min_samples_leaf=4)
modelo.fit(xtrain, ytrain)

p = modelo.predict(xtest)
accuracy_score(ytest, p)

0.8022388059701493

In [ ]:
import joblib
import os

In [ ]:
modelo.fit(x,y)
joblib.dump(modelo, os.path.join("Titanic", "model4.pkl"))